src/devices/vdml_link.c

```diff
@@ -0,0 +1,234 @@
+
+// internal function for clearing the rx buffer 
+uint32_t _clear_rx_buf(v5_smart_device_s_t* device) {
+    uint8_t buf[FIFO_SIZE];
+    return vexDeviceGenericRadioReceive(device->device_info, 
+    (uint8_t*)buf, 
+    vexDeviceGenericRadioReceiveAvail(device->device_info));
+}
+
+uint32_t link_init(uint8_t port, char* link_id, link_type_e_t type) {
+    // not used because it's a generic serial device 
+    // claim_port_i(port - 1, E_DEVICE_RADIO);
+	if (!VALIDATE_PORT_NO(port - 1)) {
+		errno = EINVAL;
+		return PROS_ERR;
+	}
+    // by default, the vexlink radio in the lower port should be by default E_DEVICE_NONE
+    // or, the only radio on the brain
+    v5_device_e_t plugged_device = registry_get_plugged_type(port - 1);
+    if(plugged_device != E_DEVICE_NONE && plugged_device != E_DEVICE_RADIO) {
+        return PROS_ERR;
+    }
+	v5_smart_device_s_t* device = registry_get_device(port - 1);
+	if (!port_mutex_take(port - 1)) {
+		errno = EACCES;
+		return PROS_ERR;
+	}
+    vexDeviceGenericRadioConnection(device->device_info, link_id, type, false);
+    return_port(port - 1, 1);
+}
+
+uint32_t link_init_override(uint8_t port, char* link_id, link_type_e_t type) {
+    // not used because it's a generic serial device 
+    // claim_port_i(port - 1, E_DEVICE_RADIO);
+	if (!VALIDATE_PORT_NO(port - 1)) {
+		errno = EINVAL;
+		return PROS_ERR;
+	}
+    // by default, the vexlink radio in the lower port should be by default E_DEVICE_NONE
+    // or, the only radio on the brain
+    v5_device_e_t plugged_device = registry_get_plugged_type(port - 1);
+    if(plugged_device != E_DEVICE_NONE && plugged_device != E_DEVICE_RADIO) {
+        return PROS_ERR;
+    }
+	v5_smart_device_s_t* device = registry_get_device(port - 1);
+	if (!port_mutex_take(port - 1)) {
+		errno = EACCES;
+		return PROS_ERR;
+	}
+    vexDeviceGenericRadioConnection(device->device_info, link_id, type, true);
+    return_port(port - 1, 1);
+}
+
+bool link_connected(uint8_t port) {
+    claim_port(port - 1, E_DEVICE_SERIAL, false);
+    bool rtv = vexDeviceGenericRadioLinkStatus(device->device_info); 
+    return_port(port - 1, rtv);
+}
+
+uint32_t link_raw_receivable_size(uint8_t port) {
+    claim_port_i(port - 1, E_DEVICE_SERIAL);
+    uint32_t rtv = vexDeviceGenericRadioReceiveAvail(device->device_info); 
+    return_port(port - 1, rtv);
+}
+
+uint32_t link_raw_transmittable_size(uint8_t port) {
+    claim_port_i(port - 1, E_DEVICE_SERIAL);
+    uint32_t rtv = vexDeviceGenericRadioWriteFree(device->device_info);
+    return_port(port - 1, rtv);
+}
+
+uint32_t link_transmit_raw(uint8_t port, void* data, uint16_t data_size) {
+    claim_port_i(port - 1, E_DEVICE_SERIAL);
+    if(!vexDeviceGenericRadioLinkStatus(device->device_info)) {
+        errno = ENXIO;
+        return_port(port - 1, PROS_ERR);
+    }
+    if(data_size > vexDeviceGenericRadioWriteFree(device->device_info)) {
+        errno = EBUSY;
+        return_port(port - 1, 0);
```

Comment:
I would, but I have a feeling a lot of use cases will probably be using pointers to GPS data structs for VEXLink so I'd rather not make it truncate data and confuse people.

Reply:
Hmm, yea for the transmit side I can see requiring space for the entire message since vexOS just drops it if there isn't space. You can't really saturate the bandwidth, so this becomes less useful.